File: 2. Add Two Numbers/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	head := l1

	for (l1 != nil && l2 != nil && l1 != l2) || (l1 != nil && l1.Val >= 10) {
		if l2 != nil && l1 != l2 {
			l1.Val += l2.Val
		}

		if l1.Val >= 10 {
			l1.Val -= 10
			if l1.Next != nil {
				l1.Next.Val += 1
			} else if l2 != nil && l2.Next != nil {
				l1.Next = l2.Next
				l1.Next.Val += 1
			} else {
				l1.Next = &ListNode{Val: 1}
			}
		}

		if l1.Next != nil {
			l1 = l1.Next
		} else {
			if l2 != nil {
				l1.Next = l2.Next
			}
			break
		}

		if l2 != nil {
			l2 = l2.Next
		}
	}

	return head
}

File: 2. Add Two Numbers/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head = new ListNode();
        ListNode* current = head;
        int carry = 0;

        while (l1 || l2 || carry != 0 ) {
            current->next = new ListNode();
            current = current->next;

            int l1val = 0;
            if (l1) {
                l1val = l1->val;
                l1 = l1->next;
            }

            int l2val = 0;
            if (l2) {
                l2val = l2->val;
                l2 = l2->next;
            }

            int sum = carry + l1val + l2val;
            if (sum < 10) {
                carry = 0;
            } else {
                carry = 1;
                sum -= 10;
            }

            current->val = sum;
        }

        return head->next;
    }
};
